application/controllers/public/Uploader.php

```diff
@@ -18,8 +18,26 @@ function __construct()
 		ini_set('memory_limit', '128M');
 
 		$allowed_groups = array(PERMISSIONS_ADMIN, PERMISSIONS_BCS, PERMISSIONS_MCS, PERMISSIONS_UPLOADER, PERMISSIONS_PLS, PERMISSIONS_READERS);
-		if (!$this->librivox_auth->has_permission($allowed_groups, $this->data['user_id']))
+		$has_group_permission = $this->librivox_auth->has_permission($allowed_groups, $this->data['user_id']);
+		if (!$has_group_permission)
```

Comment:
Thanks for getting this started! It's a relatively small change, but it will definitely improve the UX.

I have a couple of suggestions to improve this, both to organize the code, and to improve what the user sees. I would imagine it something like this:

```php
# First determine the error condition, if any.
if (!$this->ion_auth->logged_in())
{
    $ajax_error = "Your session has expired, please login again.";
    $redirect_url = "auth/login";
}
else if (!$has_group_permission)
{
    $ajax_error = "You don't have permissions for this area.";
    $redirect_url = "auth/error_no_permission";
}

# Then, based on the presence of an error condition and whether we're being called by AJAX or not,
# echo or redirect.
if (IS_AJAX and $ajax_error)
{
    echo json_encode(array(
        'error' => $ajax_error,
        # Also set the name and size, to avoid 'undefined' and 'Nan' showing up on the web page
        'name' => '',
        'size' => ''));
}

if ($redirect_url)
{
    redirect($redirect_url);
}
```